aten/src/ATen/mps/MPSAllocator.h

```diff
@@ -216,40 +244,59 @@ class MPSHeapAllocatorImpl
   // max buffer size allowed by Metal
   size_t m_max_buffer_size;
-  // sets a soft upper bound to limit the total allocations
-  bool m_set_fraction;
-  // use "PYTORCH_DEBUG_MPS_ALLOCATOR" env-var to enable debug info
-  bool m_enable_debug_info;
-
+  // maximum total size allowed to be allocated
+  size_t m_max_total_allowed_size;
+  // upper limit ratio to limit the total allocations (between 0 and 1)
+  // 0 means unlimited (would spill to disk or system failure if OOM)
+  // 1 is maximum allowed by device.recommendedMaxWorkingSetSize
+  // (e.g., value 0.95 means we allocate up to 95% of total memory; beyond that allocations fail)
+  double m_high_watermark_ratio;
+  // use "PYTORCH_DEBUG_MPS_ALLOCATOR" env-var to set debug verbosity
+  uint32_t m_debug_verbosity;
+  // default MPS stream
+  MPSStream* m_stream;
+
+  void init_allocator_params();
   HeapBlock* get_free_heap(AllocParams& p);
   bool get_free_buffer(AllocParams& p);
   BufferBlock* get_allocated_buffer_block(void* ptr);
+  BufferBlock* alloc_buffer_block(size_t size, uint32_t usage);
   bool alloc_buffer(AllocParams& p);
   void free_buffer(BufferBlock* buffer_block);
   void release_buffer(BufferBlock* buffer_block, bool remove_empty_heap = true);
   void release_buffers(BufferPool& pool);
-  bool release_available_cached_buffers(const AllocParams& p);
+  bool release_available_cached_buffers(AllocParams& p);
   bool release_cached_buffers();
-  void trigger_memory_callbacks(BufferBlock* buffer_block, IMpsAllocatorCallback::EventType event);
 
-  BufferPool& get_pool(size_t Size, bool useShared) {
-      return Size <= kMaxSmallAlloc ? (useShared ? m_small_pool_shared : m_small_pool_private) :
-                                      (useShared ? m_large_pool_shared : m_large_pool_private);
+  BufferPool& get_pool(size_t Size, uint32_t usage) {
+    if (usage & UsageFlags::SCALAR)
+      return m_scalar_pool;
+    return Size <= kMaxSmallAlloc ? ((usage & UsageFlags::SHARED) ? m_small_pool_shared : m_small_pool_private) :
+                                    ((usage & UsageFlags::SHARED) ? m_large_pool_shared : m_large_pool_private);
   }
 
-  size_t get_allocation_size(size_t Length, bool useShared) {
+  size_t get_allocation_size(size_t Length, uint32_t usage) const  {
     MTLSizeAndAlign sizeAlign = [m_device heapBufferSizeAndAlignWithLength:Length
-                                                                   options:HeapBlock::getOptions(useShared)];
+                                                                   options:HeapBlock::getOptions(usage)];
     return BufferBlock::alignUp(sizeAlign.size, sizeAlign.align);
   }
-  // TODO: make this configurable
-  static size_t max_split_size() { return std::numeric_limits<size_t>::max(); }
   // maximum size of device memory available for allocation in current process
-  size_t max_available_size() const { return [m_device recommendedMaxWorkingSetSize] - [m_device currentAllocatedSize]; }
```

Comment:
We don't need available size ?

Reply:
I separated recommendedMaxWorkingSetSize and currentAllocatedSize into two separate functions. 
Also, this max_available_size() function was wrong since the currentAllocatedSize could go larger than recommendedMaxWorkingSetSize. The max_available_size() was not used anywhere, anyway.